pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java

```diff
@@ -171,7 +172,14 @@ public ValidDocIndexReader getValidDocIndex() {
   @Override
   public GenericRow getRecord(int docId, GenericRow reuse) {
     // NOTE: Use PinotSegmentRecordReader to read immutable segment
-    throw new UnsupportedOperationException();
+    try {
+      PinotSegmentRecordReader pinotSegmentRecordReader = new PinotSegmentRecordReader(this);
```

Comment:
I don't think we should instantiate a record reader from the segment and too per call